browserdebuggertools/chrome/interface.py

```diff
@@ -185,9 +182,121 @@ def get_opened_javascript_dialog(self):
             self._socket_handler.event_handlers["JavascriptDialog"].get_opened_javascript_dialog()
         )
+
+    def get_iframe_source_content(self, xpath):
+        # type: (str) -> str
+        """
+        Returns the HTML markup for an iframe document, where the iframe node can be located in the
+        DOM with the given xpath.
+
+        :param xpath: following the spec 3.1 https://www.w3.org/TR/xpath-31/
+        :return: HTML markup
+        :raises XPathSyntaxError: The given xpath is invalid
+        :raises IFrameNotFoundError: A matching iframe document could not be found
+        :raises UnknownError: The socket handler received a message with an unknown error code
+        """
+
+        try:
+            XPath(xpath)  # Validates the xpath
+
+        except XPathSyntaxError:
+            raise InvalidXPathError("{0} is not a valid xpath".format(xpath))
+
+        return self._get_iframe_html(xpath)
+
+    def get_page_source(self):
+        # type: () -> str
+        """
+        Returns the HTML markup of the current page. Iframe tags are included but the enclosed
+        documents are not. Consider enabling the Page domain to increase performance.
+
+        :return: HTML markup
+        """
+
+        root_node_id = self._socket_handler.event_handlers["PageLoad"].get_root_node_id()
+        return self._get_outer_html(root_node_id)
+
+    def _get_iframe_html(self, xpath, _attempts=0):
+        # type: (str, int) -> str
+
+        try:
+            backend_node_id = self._get_iframe_backend_node_id(xpath)
+            return self._get_outer_html(backend_node_id)
+
+        except ResourceNotFoundError:
+            # The cached node doesn't exist any more, so we need to find a new one that matches
+            # the xpath. Backend node IDs are unique, so there is not a risk of getting the
+            # outer html of the wrong node.
+            if _attempts < 1:
+                if xpath in self._node_map:
+                    del self._node_map[xpath]
+                return self._get_iframe_html(xpath, _attempts=_attempts + 1)
+
+            raise IFrameNotFoundError()
+
+    def _get_iframe_backend_node_id(self, xpath):
+        # type: (str) -> int
+
+        if xpath in self._node_map:
+            return self._node_map[xpath]
+
+        try:
+            node_info = self._get_info_for_first_matching_node(xpath)
+            backend_node_id = node_info["node"]["contentDocument"]["backendNodeId"]
+        except (KeyError, ResourceNotFoundError):
+            raise IFrameNotFoundError()
+
+        self._node_map[xpath] = backend_node_id
+        return backend_node_id
+
+    def _get_info_for_first_matching_node(self, xpath):
+        # type: (str) -> dict
+
+        search_info = self._perform_search(xpath)
+        if search_info.get("resultCount", 0) > 0:
+            search_results = self._get_search_results(search_info["searchId"], 0, 1)
+            self._discard_search(search_info["searchId"])
```

Comment:
It's likely that `_perform_search()`, `_get_search_results()` and `_discard_search()` will always get called together. Meaning they could be in one general search function otherwise we could end up with the case where `_perform_search()` is called and we lose the searchIds which can potentially introduce memory leak.
Alternatively after every search we cache the searchId and remove all the search results upon calling `reset()` but this is less ideal.

Reply:
I created a context manager _get_node_ids which can be used to return up to a specified maximum number of nodeIds that match a given xpath, and always ensures the search session is cleared up.